libs/s25main/mapGenerator/Algorithms.h

```diff
@@ -0,0 +1,298 @@
+     * the area are set to a maximum value of width + height of the map.
+     *
+     * @param size size of  the map
+     * @param area area to compute distance values for
+     * @param defaultValue default distance value applied to points outside of the specified area
+     * @param evaluator evaluator function takes a MapPoint as input and returns `true` or `false`
+     *
+     * @return distance of each grid position to closest point which has been evaluted with `true`.
+     */
+    template<typename T, class T_Container>
+    ValueMap<unsigned> Distances(const MapExtent& size, const T_Container& area, const unsigned defaultValue,
+                                 T&& evaluator)
+    {
+        const unsigned maximumDistance = size.x * size.y;
+
+        std::queue<MapPoint> queue;
+        ValueMap<unsigned> distances(size, defaultValue);
+
+        for(const MapPoint& pt : area)
+        {
+            if(evaluator(pt))
+            {
+                distances[pt] = 0;
+                queue.push(pt);
+            } else
+            {
+                distances[pt] = maximumDistance;
+            }
+        }
+
+        UpdateDistances(distances, queue);
+
+        return distances;
+    }
+
+    /**
+     * Computes a map of distance values describing the distance of each grid position to the closest position for
+     * which the evaluator returned `true`.
+     *
+     * @param size size of  the map
+     * @param evaluator evaluator function takes a MapPoint as input and returns `true` or `false`
+     *
+     * @return distance of each grid position to closest point which has been evaluted with `true`.
+     */
+    template<typename T_Value>
+    ValueMap<unsigned> Distances(const MapExtent& size, T_Value&& evaluator)
+    {
+        const unsigned maximumDistance = size.x * size.y;
+
+        std::queue<MapPoint> queue;
+        ValueMap<unsigned> distances(size, maximumDistance);
+
+        RTTR_FOREACH_PT(MapPoint, size)
+        {
+            if(evaluator(pt))
+            {
+                distances[pt] = 0;
+                queue.push(pt);
+            }
+        }
+
+        UpdateDistances(distances, queue);
+
+        return distances;
+    }
+
+    /**
+     * Computes an upper limit for the specified values. The number of values between the specified minimum and the
+     * computed limit is at least as high as the specified coverage of the map.
+     *
+     * @param values map of comparable values
+     * @param area area of nodes to consider (empty for the whole map)
+     * @param coverage percentage of expected map coverage (value between 0 and 1)
+     * @param minimum minimum value to consider
+     *
+     * @returns a value between the specified minimum and the maximum value of the map.
+     */
+    template<typename T, class T_Area>
+    unsigned Count(const ValueMap<T>& values, const T_Area& area, T minimum, T maximum)
+    {
+        unsigned valuesInRange = 0;
+
+        if(area.empty())
```

Comment:
I'd not use that. Use 2 functions because the count of an empty are being equal to the count of the whole map feels wrong. And as you essentially switch algorithms, making it 2 functions (one not taking an area) makes more sense. This tickles further so the `WholeMap` function which returns an empty vector can go too

Reply:
Fixed (removed "WholeMap" and split up into separate functions)